File: 211/main.py
"""
Design a data structure that supports the following two operations:

void addWord(word)
bool search(word)

search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.

For example:

addWord("bad")
addWord("dad")
addWord("mad")
search("pad") -> false
search("bad") -> true
search(".ad") -> true
search("b..") -> true

Note:
You may assume that all words are consist of lowercase letters a-z.

click to show hint.
You should be familiar with how a Trie works. If not, please work on this problem: Implement Trie (Prefix Tree) first.
"""

class TrieNode(object):
    def __init__(self, if_terminal = False):
        """
        Initialize your data structure here.
        """
        self.children = {}
        self.terminal = if_terminal

class WordDictionary(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        curr = self.root
        for ch in word:
            if ch not in curr.children:
                curr.children[ch] = TrieNode()
            curr = curr.children[ch]
        curr.terminal = True

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        def searchRecur(node, s):
            if len(s) == 0:
                return node.terminal
            else:
                ch = s[0]
                ans = False
                if ch == '.':
                    for child in node.children.values():
                        ans |= searchRecur(child, s[1:])
                elif ch not in node.children:
                    ans = False
                else:
                    ans = searchRecur(node.children[ch], s[1:])
                return ans

        return searchRecur(self.root, word)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
trie = WordDictionary()
trie.addWord("bad")
trie.addWord("dad")
trie.addWord("mad")
print(trie.search("pad"))
print(trie.search("bad"))
print(trie.search(".ad"))
print(trie.search("b.."))

File: 211/main.ts
class WordDictionary {
  terminal: boolean;
  children: Map<string, WordDictionary>;
  constructor() {
    this.children = new Map();
    this.terminal = false;
  }

  addWord(word: string): void {
    if (word.length === 0) {
      this.terminal = true;
      return;
    }
    const firstChar = word[0];
    if (!this.children.has(firstChar)) {
      this.children.set(firstChar, new WordDictionary());
    }
    this.children.get(firstChar)?.addWord(word.slice(1));
  }

  search(word: string): boolean {
    if (word.length === 0) return this.terminal;
    const firstChar = word[0];
    if (firstChar === ".") {
      return Array.from(this.children.values())
        .map((value) => value.search(word.slice(1)))
        .some((b) => b);
    }
    if (!this.children.has(firstChar)) return false;
    return this.children.get(firstChar)!.search(word.slice(1));
  }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */

let wordDictionary = new WordDictionary();
wordDictionary.addWord("bad");
wordDictionary.addWord("dad");
wordDictionary.addWord("mad");
console.log(wordDictionary.search("pad")); // return False
console.log(wordDictionary.search("bad")); // return True
console.log(wordDictionary.search(".ad")); // return True
console.log(wordDictionary.search("b.."));
